torch/utils/_sympy/functions.py

```diff
@@ -1,43 +1,78 @@
+import functools
 import math
+import sys
 
 import sympy
 from sympy import S
-from sympy.core.logic import fuzzy_and, fuzzy_not, fuzzy_or
 
 __all__ = [
     "FloorDiv",
     "ModularIndexing",
     "CleanDiv",
     "CeilDiv",
-    "Pow",
-    "TrueDiv",
+    "IntTrueDiv",
+    "FloatTrueDiv",
     "LShift",
     "RShift",
     "IsNonOverlappingAndDenseIndicator",
-    "Round",
+    "RoundToInt",
     "RoundDecimal",
+    "ToFloat",
+    "FloatPow",
+    "PowByNatural",
 ]
 
 
+def _keep_float(f):
+    @functools.wraps(f)
+    def inner(*args):
+        r = f(*args)
+        if any(isinstance(a, sympy.Float) for a in args) and not isinstance(
+            r, sympy.Float
+        ):
+            r = sympy.Float(float(r))
+        return r
+
+    return inner
+
+
 def fuzzy_eq(x, y):
     if None in (x, y):
         return None
     return x == y
 
 
+# It would be nice to have assertions on whether or not inputs is_integer
+# However, with bugs like https://github.com/sympy/sympy/issues/26620 sympy
```

Comment:
It wasn't obvious to me how to "special case zero" here, because if we for example do `0.0 * 0.0` and this becomes `0` (which is a simplification Sympy does right now), I have now lost the floatness of this input and I may propagate incorrect types unpredictably to the point where I have a chance to special case it.

Reply:
inb4, we'll end up having strongly typed SymPy expressions all across, similar to how we do in `index_propagation.py`

Can you elaborate on the `0.0 * 0.0 = 0` issue? How is this an issue when you have the keep floatness wrapper.

Reply:
I would really like to have strongly typed sympy expressions, but I got stuck fixing all the sites in Inductor, there are still a bunch of places we play fast and loose and I couldn't figure them all out.

The keep float-ness wrapper only applies when you are doing value ranges. If someone does some Sympy operations manually, you can still lose it. I am kind of wondering if we  could potentially fix this using the sympy multiple dispatch API, not sure...